keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,366 @@
+| Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) | High | This API should serve only complete information about workloads. If kubelet is in the init phase and not all pod’s information is known, then the API should report that information. |
+| Kube-API is down or unreachable | Low | Kube-API availability should not impact this API. When the control plane is down or unreachable but kubelet is working properly this API should return most recent data about local pods that are computed by kubelet even if those data were not reported or accepted by Kube-API. |
+
+#### Control Plane availability issue
+
+The Kubernetes Control Plane is the source of truth and guarantees system integrity. However, already running pods should not be dependent on Control Plane availability or connectivity issues. When Node is healthy but has problems accessing Kube-API running pods, statuses should be refreshed and stored as usual.
+The new API should return pods statuses that are computed by kubelet even though they are not reported back to kube-API but reflect the current pod’s state.
+
+#### Kubelet restarts issue
+
+Kubelet restart [issue](https://github.com/kubernetes/kubernetes/issues/100277) is impacting this API. After restart kubelet marks all pods as NotReady and then starts the init loop checking the actual pod's status. When kubelet is in this `unknown` state it should not serve the API data which might be incorrect. This API needs to be reliable and returned data should reflect the actual state of the pods. We don't want this API to return partial data. We should resolve this issue before the beta release. One of the possible fixes is returning error code indicating that kubelet is in the init phase and that pod's data are not known yet.
+
+## Design Details
+
+### Proposed API
+
+We propose to add new gPRC API `podinfo` in Kubelet, listening on a unix socket at `/var/lib/kubelet/pod-info/kubelet.sock`. The endpoint will be versioned.
+The gRPC Service will expose 3 methods serving local pods statuses data:
+
+```protobuf
+service PodInfo {
+    // ListPodConditions returns a of List of PodConditions
+    rpc ListPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // WatchPodConditions returns a stream of List of PodConditions
+    // Whenever a pod state change api returns the new list
+    rpc WatchPodConditions(Empty) returns (stream PodConditionsResponse) {}
+    // GetPodConditions returns a PodInfo for given pod's UID
+    rpc GetPodConditions(GetPodConditionsRequest) returns (PodInfo) {}
+}
+
+// Conditions aligns with v1.PodCondition
+message Conditions {
+      string Type = 1;
+      string Status = 2;
+      Timestamp LastProbeTime = 3;
+      Timestamp LastTransitionTime = 4;
+      string Reason = 5;
+      string Message = 6;
+}
+
+// PodInfo returns a Pod details and list of status Conditions
+message PodInfo {
```

Comment:
PodStatus has last message and last reason as a top level values. Do you want to do the same?

Reply:
no I don't think it is needed.

Reply:
so the way to get those would be to check sort conditions by timestamp? Or you are saying this information is not needed?

Reply:
I think it is not needed. We have max 5 conditions and every condition occurs once in the list so based on that condition the workload can compute the readiness status.
I other work if `Ready` condition is set we don't really care if `ContainerReady` was set first. And if `Ready` is false or not present we just check the `ContainerReady` and we don't care about the timestamps. Right?